apps/nsq_tail/nsq_tail.go

```diff
@@ -91,29 +100,42 @@ func main() {
 	cfg.UserAgent = fmt.Sprintf("nsq_tail/%s go-nsq/%s", version.Binary, nsq.VERSION)
 	cfg.MaxInFlight = *maxInFlight
 
-	consumer, err := nsq.NewConsumer(*topic, *channel, cfg)
-	if err != nil {
-		log.Fatal(err)
-	}
+	consumers := []nsq.Consumer{}
+	for i := 0; i < len(topics); i += 1 {
+		fmt.Printf("Adding consumer for topic: %s\n", topics[i])
 
-	consumer.AddHandler(&TailHandler{totalMessages: *totalMessages})
+		consumer, err := nsq.NewConsumer(topics[i], *channel, cfg)
+		if err != nil {
+			log.Fatal(err)
+		}
 
-	err = consumer.ConnectToNSQDs(nsqdTCPAddrs)
-	if err != nil {
-		log.Fatal(err)
-	}
+		consumer.AddHandler(&TailHandler{topicName: topics[i], totalMessages: *totalMessages})
 
-	err = consumer.ConnectToNSQLookupds(lookupdHTTPAddrs)
-	if err != nil {
-		log.Fatal(err)
+		err = consumer.ConnectToNSQDs(nsqdTCPAddrs)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		err = consumer.ConnectToNSQLookupds(lookupdHTTPAddrs)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		consumers = append(consumers, *consumer)
 	}
 
-	for {
+	for _, consumer := range consumers {
 		select {
 		case <-consumer.StopChan:
```

Comment:
This is not really useful: it will only wait for the first consumer to stop.

I don't think any consumers will stop before we tell them to though, they will try to reconnect. So this case could just be removed (and the loop as well). This just needs to wait for the sigChan and then tell all consumers to stop (as you have below). After that, to be most correct, it should then wait for all consumers to have stopped.

Reply:
@ploxiln I've simply rewrited existing logic. Are you sure, that we need to have only waiting for `SigChan`?

Reply:
It wasn't 100% correct before, but in the process of rewriting it you made it more nonsensical. The previous bare for loop is different in nature from a loop over all consumers.